pkg/store/store.go

```diff
@@ -296,30 +300,59 @@ func (t *Store) Count(prefix schema.KeyPrefix) (count *schema.ItemsCount, err er
 	return
 }
 
-// fixme(leogr): need to be optmized
-func (t *Store) itemAt(readTs uint64) (version uint64, key, value []byte, err error) {
-	stream := t.db.NewStreamAt(readTs)
-	stream.ChooseKey = func(item *badger.Item) bool {
-		return item.UserMeta() != bitTreeEntry && item.Version() == readTs
+func (t *Store) itemAt(readTs uint64) (index uint64, key, value []byte, err error) {
+	if readTs > t.tree.w {
+		return 0, nil, nil, errors.New("Index not found")
 	}
+	index = readTs
 
-	found := false
-
-	stream.Send = func(list *pb.KVList) error {
-		for _, kv := range list.Kv {
-			key = kv.Key
-			value = kv.Value
-			version = kv.Version
-			found = true
+	var refkey []byte
+	// cache reference lookup
+	if key := t.tree.rcaches[0].Get(readTs - 1); key != nil {
```

Comment:
`t.tree` should be read-locked here, shouldn't?

Reply:
Tree is now locked